File: obligation-cordapp/java-source/src/test/java/net/corda/examples/obligation/flows/TransferObligationTests.java
package net.corda.examples.obligation.flows;

import net.corda.core.transactions.SignedTransaction;
import net.corda.examples.obligation.Obligation;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import static net.corda.finance.Currencies.POUNDS;
import static net.corda.testing.internal.InternalTestUtilsKt.chooseIdentity;
import static org.hamcrest.CoreMatchers.instanceOf;
import static org.junit.Assert.assertEquals;

public class TransferObligationTests extends ObligationTests {

    @Rule
    public final ExpectedException exception = ExpectedException.none();

    @Test
    public void transferNonAnonymousObligationSuccessfully() throws Exception {
        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Transfer obligation.
        SignedTransaction transferTransaction = transferObligation(issuedObligation.getLinearId(), b, c, false);
        network.waitQuiescent();
        Obligation transferredObligation = (Obligation) transferTransaction.getTx().getOutputStates().get(0);

        // Check the issued obligation with the new lender is the transferred obligation
        assertEquals(issuedObligation.withNewLender(chooseIdentity(c.getInfo())), transferredObligation);

        // Check everyone has the transfer transaction.
        Obligation aObligation = (Obligation) a.getServices().loadState(transferTransaction.getTx().outRef(0).getRef()).getData();
        Obligation bObligation = (Obligation) b.getServices().loadState(transferTransaction.getTx().outRef(0).getRef()).getData();
        Obligation cObligation = (Obligation) c.getServices().loadState(transferTransaction.getTx().outRef(0).getRef()).getData();
        assertEquals(aObligation, bObligation);
        assertEquals(bObligation, cObligation);
    }

    @Test
    public void transferAnonymousObligationSuccessfully() throws Exception {
        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Transfer obligation.
        SignedTransaction transferTransaction = transferObligation(issuedObligation.getLinearId(), b, c, true);
        network.waitQuiescent();
        Obligation transferredObligation = (Obligation) transferTransaction.getTx().getOutputStates().get(0);

        // Check the issued obligation with the new lender is the transferred obligation.
        assertEquals(issuedObligation.withNewLender(transferredObligation.getLender()), transferredObligation);

        // Check everyone has the transfer transaction.
        Obligation aObligation = (Obligation) a.getServices().loadState(transferTransaction.getTx().outRef(0).getRef()).getData();
        Obligation bObligation = (Obligation) b.getServices().loadState(transferTransaction.getTx().outRef(0).getRef()).getData();
        Obligation cObligation = (Obligation) c.getServices().loadState(transferTransaction.getTx().outRef(0).getRef()).getData();
        assertEquals(aObligation, bObligation);
        assertEquals(bObligation, cObligation);
    }

    @Test
    public void transferFlowCanOnlyBeStartedByLender() throws Exception {
        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Transfer obligation.
        exception.expectCause(instanceOf(IllegalStateException.class));
        transferObligation(issuedObligation.getLinearId(), a, c, false);
    }
}

File: cordapp-upgrades/v1-initial/src/test/kotlin/net/corda/examples/obligation/flows/TransferObligationTests.kt
package net.corda.examples.obligation.flows

import net.corda.examples.obligation.Obligation
import net.corda.finance.POUNDS
import net.corda.testing.internal.chooseIdentity
import org.jgroups.util.Util.assertEquals

class TransferObligationTests : ObligationTests() {

    @org.junit.Test
    fun `Transfer non-anonymous obligation successfully`() {
        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Transfer obligation.
        val transferTransaction = transferObligation(issuedObligation.linearId, b, c, anonymous = false)
        network.waitQuiescent()
        val transferredObligation = transferTransaction.tx.outputStates.first() as Obligation

        // Check the issued obligation with the new lender is the transferred obligation
        assertEquals(issuedObligation.withNewLender(c.info.chooseIdentity()), transferredObligation)

        // Check everyone has the transfer transaction.
        val aObligation = a.services.loadState(transferTransaction.tx.outRef<Obligation>(0).ref).data as Obligation
        val bObligation = b.services.loadState(transferTransaction.tx.outRef<Obligation>(0).ref).data as Obligation
        val cObligation = c.services.loadState(transferTransaction.tx.outRef<Obligation>(0).ref).data as Obligation
        assertEquals(aObligation, bObligation)
        assertEquals(bObligation, cObligation)
    }

    @org.junit.Test
    fun `Transfer anonymous obligation successfully`() {
        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Transfer obligation.
        val transferTransaction = transferObligation(issuedObligation.linearId, b, c)
        network.waitQuiescent()
        val transferredObligation = transferTransaction.tx.outputStates.first() as Obligation

        // Check the issued obligation with the new lender is the transferred obligation.
        assertEquals(issuedObligation.withNewLender(transferredObligation.lender), transferredObligation)

        // Check everyone has the transfer transaction.
        val aObligation = a.services.loadState(transferTransaction.tx.outRef<Obligation>(0).ref).data as Obligation
        val bObligation = b.services.loadState(transferTransaction.tx.outRef<Obligation>(0).ref).data as Obligation
        val cObligation = c.services.loadState(transferTransaction.tx.outRef<Obligation>(0).ref).data as Obligation
        assertEquals(aObligation, bObligation)
        assertEquals(bObligation, cObligation)

    }

    @org.junit.Test
    fun `Transfer flow can only be started by lender`() {
        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Transfer obligation.
        kotlin.test.assertFailsWith<IllegalStateException> {
            transferObligation(issuedObligation.linearId, a, c, anonymous = false)
        }
    }

}
